writers_rollingfilewriter.go

```diff
@@ -138,7 +140,7 @@ var compressionTypes = map[rollingArchiveType]compressionType{
 			if err != nil {
 				return err
 			}
-			return createGzip(archiveName, tar)
+			return createGzip(archiveName, bytes.NewReader(tar))
```

Comment:
So, this is an interested path. The reason that I didn't convert everything to Readers is because it would require more in-depth plumbing than I'm familiar with given my lack of Go experience. What you're saying makes sense about giving createTar the Writer to write to. My opinion is that the root problem is that we always want to Pipe (as you noted) in this case, as we never call `unarchive` unless it is immediately followed by an `archive` call. So really the `archive` vs. `unarchive` primitives in the original code are maybe mis-designed ... what we really want is `newArchive` vs. `addFilesToArchive` -- if we exposed those higher-level functions then the `addFilesToArchive` could control the "unarchive-and-rearchive" process itself rather than trying to make the unarchiving its own first-class concept. Does that make sense to you?